File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Animations/VFXAction.ts
import Async from "../../../../../Core/Async";
import DungeonManager from "../../DungeonManager";
import { AbstractEffect } from "../../Effect/AbstractEffect";
import { EffectOrder } from "../../Effect/EffectKit";
import AbstractGameAction from "../AbstractGameAction";

export default class VFXAction extends AbstractGameAction
{
    constructor(private Effect: AbstractEffect, private Interval:number = 0,private Order = EffectOrder.MID) {
        super();
    }

    async Update(): Promise<boolean> {
        DungeonManager.EffectManager.Play(this.Effect);
        await Async.Delay(this.Interval);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Utility/VFXAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Dungeon;
using ModTheSpire.Event;

namespace ModTheSpire.Effect.Other
{
    public class VFXAction : IGameAction
    {
        private readonly AbstractEffect mEffect;
        private readonly int mInterval;
        private readonly EffectOrder mOrder;
        public float Duration { get; set; }

        public ActionType Type { get; } = ActionType.WAIT;
        public VFXAction(AbstractEffect effect, int interval, EffectOrder order = EffectOrder.MID)
        {
            mEffect = effect;
            mInterval = interval;
            mOrder = order;
        }
        
        public void Init()
        {
        }

        public async UniTask<bool> Update()
        {
            DungeonManager.EffectManager.Play(mEffect);
            await UniTask.Delay(this.mInterval);
            return true;
        }
    }
}